scipy/optimize/_milp.py

```diff
@@ -0,0 +1,359 @@
+           "HiGHS - high performance software for linear optimization."
+           Accessed 12/25/2021 at https://www.maths.ed.ac.uk/hall/HiGHS/#guide
+    .. [2] Huangfu, Q. and Hall, J. A. J. "Parallelizing the dual revised
+           simplex method." Mathematical Programming Computation, 10 (1),
+           119-142, 2018. DOI: 10.1007/s12532-017-0130-5
+
+    Examples
+    --------
+    Consider the problem at
+    https://en.wikipedia.org/wiki/Integer_programming#Example, which is
+    expressed as a maximization problem of two variables. Since `milp` requires
+    that the problem be expressed as a minimization problem, the objective
+    function coefficients on the decision variables are:
+
+    >>> c = -np.array([0, 1])
+
+    Note the negative sign: we maximize the original objective function
+    by minimizing the negative of the objective function.
+
+    We collect the coefficients of the constraints into arrays like:
+
+    >>> A = np.array([[-1, 1], [3, 2], [2, 3]])
+    >>> b_u = np.array([1, 12, 12])
+    >>> b_l = np.full_like(b_u, -np.inf)
+
+    Because there is no lower limit on these constraints, we have defined a
+    variable ``b_l`` full of values representing negative infinity. This may
+    be unfamiliar to users of `scipy.optimize.linprog`, which only accepts
+    "less than" (or "upper bound") inequality constraints of the form
+    ``A_ub @ x <= b_u``. By accepting both ``b_l`` and ``b_u`` of constraints
+    ``b_l <= A_ub @ x <= b_u``, `milp` makes it easy to specify "greater than"
+    inequality constraints, "less than" inequality constraints, and equality
+    constraints concisely.
+
+    These arrays are collected into a single `LinearConstraint` object like:
+
+    >>> from scipy.optimize import LinearConstraint
+    >>> constraints = LinearConstraint(A, b_l, b_u)
+
+    The non-negativity bounds on the decision variables are enforced by
+    default, so we do not need to provide an argument for `bounds`.
+
+    Finally, the problem states that both decision variables must be integers:
+
+    >>> integrality = np.ones_like(c)
+
+    We solve the problem like:
+    >>> from scipy.optimize import milp
+    >>> res = milp(c=c, constraints=constraints, integrality=integrality)
+    >>> res.x
+    [1.0, 2.0]
+
+    Note that had we solved the relaxed problem (without integrality
+    constraints):
+    >>> res = milp(c=c, constraints=constraints)  # OR:
+    >>> # from scipy.optimize import linprog; res = linprog(c, A, b_u)
+    >>> res.x
+    [1.8, 2.8]
+
+    we would not have obtained the correct solution by rounding to the nearest
+    integers.
+
+    """
+
+    args_iv = _milp_iv(c, integrality, bounds, constraints, options)
+    c, integrality, lb, ub, indptr, indices, data, b_l, b_u, options = args_iv
+
+    highs_res = _highs_wrapper(c, indptr, indices, data, b_l, b_u,
+                               lb, ub, integrality, options)
+
+    res = {}
+
+    status_map = {7: 0, 13: 1, 8: 2, 10: 3}
```

Comment:
These look a little "magic"-y to me.  Would prefer the same strategy as in `scipy.optimize._linprog_highs` that uses named constants or a refactor to have both of those functions use the same mapping dictionary

Reply:
ok i'll take a look at how it's done there.

Reply:
There is a dictionary in `_linprog_highs.py` that can be moved out of a function so we can import it here. I should do that in a separate PR. 

<details>

```python3
# conversion between HiGHS and SciPy status number/message
highs_statuses = {
    MODEL_STATUS_NOTSET: (
        4,
        f'HiGHS Status Code {MODEL_STATUS_NOTSET}: HighsModelStatusNOTSET',
    ),
    MODEL_STATUS_LOAD_ERROR: (
        4,
        f'HiGHS Status Code {MODEL_STATUS_LOAD_ERROR}: '
        'HighsModelStatusLOAD_ERROR',
    ),
    MODEL_STATUS_MODEL_ERROR: (
        2,
        f'HiGHS Status Code {MODEL_STATUS_MODEL_ERROR}: '
        'HighsModelStatusMODEL_ERROR',
    ),
    MODEL_STATUS_PRESOLVE_ERROR: (
        4,
        f'HiGHS Status Code {MODEL_STATUS_PRESOLVE_ERROR}: '
        'HighsModelStatusPRESOLVE_ERROR',
    ),
    MODEL_STATUS_SOLVE_ERROR: (
        4,
        f'HiGHS Status Code {MODEL_STATUS_SOLVE_ERROR}: '
        'HighsModelStatusSOLVE_ERROR',
    ),
    MODEL_STATUS_POSTSOLVE_ERROR: (
        4,
        f'HiGHS Status Code {MODEL_STATUS_POSTSOLVE_ERROR}: '
        'HighsModelStatusPOSTSOLVE_ERROR',
    ),
    MODEL_STATUS_MODEL_EMPTY: (
        4,
        f'HiGHS Status Code {MODEL_STATUS_MODEL_EMPTY}: '
        'HighsModelStatusMODEL_EMPTY',
    ),
    MODEL_STATUS_RDOVUB: (
        4,
        f'HiGHS Status Code {MODEL_STATUS_RDOVUB}: '
        'HighsModelStatusREACHED_DUAL_OBJECTIVE_VALUE_UPPER_BOUND',
    ),
    MODEL_STATUS_REACHED_OBJECTIVE_TARGET: (
        4,
        f'HiGHS Status Code {MODEL_STATUS_REACHED_OBJECTIVE_TARGET}: '
        'HighsModelStatusREACHED_OBJECTIVE_TARGET',
    ),
    MODEL_STATUS_INFEASIBLE: (
        2,
        "The problem is infeasible.",
    ),
    MODEL_STATUS_UNBOUNDED: (
        3,
        "The problem is unbounded.",
    ),
    MODEL_STATUS_UNBOUNDED_OR_INFEASIBLE: (
        4,
        "The problem is unbounded or infeasible.",
    ),
    MODEL_STATUS_OPTIMAL: (
        0,
        "Optimization terminated successfully.",
    ),
    MODEL_STATUS_REACHED_TIME_LIMIT: (
        1,
        "Time limit reached.",
    ),
    MODEL_STATUS_REACHED_ITERATION_LIMIT: (
        1,
        "Iteration limit reached.",
    ),
}
```

</details>

While we're at it, some of those messages like `f'HiGHS Status Code {MODEL_STATUS_REACHED_OBJECTIVE_TARGET}: '
        'HighsModelStatusREACHED_OBJECTIVE_TARGET'` are not very informative. Should we let the HiGHS message pass through directly for those? 

Do we still think we should replace the HiGHS messages with the SciPy messages?

Or how about in every message we concatenate the HIGHS status/message (after the usual SciPy message) in parenetheses, e.g. f"HiGHS Status Code {highs_status}: {highs_message}"?

Reply:
> Or how about in every message we concatenate the HIGHS status/message (after the usual SciPy message) in parenetheses, e.g. f"HiGHS Status Code {highs_status}: {highs_message}"?

Yes, I agree.  This looks like a good way to go

Reply:
Done.